### ISSUE TITLE
NA returned with Warning: Embedding 8 connected components using meta-embedding (experimental)   n_components

### ISSUE THREAD

**Issue by YubinXie** [lmcinnes/umap#90]
Working enviroment: Mac OS 13.5, PYTHON 3.6.
My data is 660K * 6 dimensions. Firstly, I tried n_neighbors =100. It worked fine. Then I tried n_neighbors=15, it gave warning:

`lib/python3.6/site-packages/umap/spectral.py:229: UserWarning: Embedding 8 connected components using meta-embedding (experimental) n_components`

And the returned embedding is all NA. Then I also tried  n_neighbors =200,500, all the embedding is NA. I am not sure what happened. 

Thank you!

**Comment by lmcinnes** [comment]
I suspect the spectral initialisation is failing for one reason or another.
This can often happen for particularly oddly distributed data. As a
workaround you can use ``init='random'`` as a parameter to UMAP. It should
stop the NaNs happening at least. This isn't ideal, but it should get you
past the immediate problem. I'll try to look into the deeper issue soon.

**Comment by YubinXie** [comment]
Thank you for your quick reply and insightful suggestion. Yes, my data distribution is odd (very spare and low dimension). But any idea that why it worked before, and all of a sudden, it no longer works? Is different run of UMAP affect each other? Thanks! (UMAP is a great visualization tool!)

**Comment by lmcinnes** [comment]
The nearest neighbor computation is approximate and somewhat stochastic, so
you can get differences. More likely 100 was a large enough value that it
didn't shatter things to badly and break the spectral initialisation --
larger values may have been large enough to create a very small eigengap
and break things in other ways. It is odd that it was that sensitive. If it
is at all possible to share your data I would be interested to experiment
and try to reproduce the problem -- but I certainly understand if you can't.

**Comment by lmcinnes** [comment]
There's definitely a bug in there then -- unfortunately one I can't
reproduce right now. Thanks for the report though: I'll be on the lookout
for anytign similar and hopefully I can track it down.

**Comment by YubinXie** [comment]
Some update: 
n_neighbors =100 used to work. But when I restarted my jupyter notebook, n_neighbors =100 also returned NA. I think the default random seed should be the same. Not sure why this is happening.

**Comment by lmcinnes** [comment]
This is certainly odd. I'll have to see if I can manage to reproduce the
problem to try and track down what is going astray.

**Comment by JoshuaC3** [comment]
I found this issue (Embedding 6 connected instead), and another issue, 

```
"Random Projection forest initialisation failed due to recursion"
            "limit being reached. Something is a little strange with your "
            "data, and this may take longer than normal to compute."
```

when trying to embed weekday data. Hence, I had only 7 unique rows in my np.array and, after much spluttering and wasted time I got the following plot,

![weekday_embedding](https://user-images.githubusercontent.com/11645712/43518194-2d1d86ea-9583-11e8-83a6-0e77845c44a7.png)
(Interesting how it embeds each instance of the same day of the week in a sightly different (x, y) co-ordinate... @lmcinnes any intuition on this?)

My suggestion would be a check on the number of unique rows in the np.array a users provides. Still, I am not 100% certain that this was the cause of my two warnings but hopefully it might help.

**Comment by lmcinnes** [comment]
Each instance is viewed as an independent object, so even if they are
identical in the data they are treated as technically separate, and thus
embed into different locations.

Checking for unique rows is certainly an option, but a very expensive one
computationally. There are other checks that should catch such situations,
so I'm not sure whether this was technically the problem or not.

**Comment by ahartikainen** [comment]
I have had this problem too. I transformed some image data to (n,m), and it happened when n was large.

Maybe there is some underflow/overflow/negative-log-sqrt somewhere?

**Comment by lmcinnes** [comment]
It could be -- as far as I can tell it is happening in sklearn nor umap --
so it remains to track down which sklearn calls are at issue (we don't have
a traceback). I'll continue looking for a consistent reproducer.

**Comment by khamkarajinkya** [comment]
Running into similar issues

parameters  = `n_neighbors = 50, min_dist = 0.05, metric = 'euclidean' ` 

Datatset: 100000 vectors with 300 dimensions

**Comment by proudquartz** [comment]
Awesome work. I am having the same issue here as well.

Dataset: 102300 vectors with 95 features.

Default parameters produces very nice clusters. But when I increase n_neighbors, the embedding becomes NA.

parameters: n_neighbors = 50, everything else is default

**Comment by lmcinnes** [comment]
The most recent versions on github should have fixed this -- are you
running against that, or something from pip or conda?

**Comment by bioguy2018** [comment]
Hi I also have faced this problem but I manged to solve it with init=random
but is this something that I should worry about in general?

**Comment by lmcinnes** [comment]
I think it is fixed, and I should push out a patch release -- thanks for the reminder.

**Comment by bioguy2018** [comment]
@lmcinnes  Thanks a lot just to mention in my case I didn't have any NaN problem but rather this

> UserWarning: Embedding a total of 2 separate connected components using meta-embedding (experimental)
  n_components
but as long as I understood its not an issue right ? Looking forward for the update :) 
Thanks a lot again

**Comment by lmcinnes** [comment]
Yes -- that was an experimental feature in 0.3, but it seems to be working well, so you can ignore it.

**Comment by sleighsoft** [comment]
This might be related: https://github.com/scikit-learn/scikit-learn/issues/13393
